lib/Frontend/ParseableInterfaceSupport.cpp

```diff
@@ -400,6 +401,134 @@ static void printImports(raw_ostream &out, ModuleDecl *M) {
+static bool isPublicOrUsableFromInline(const ValueDecl *VD) {
+  AccessScope scope =
+      VD->getFormalAccessScope(/*useDC*/nullptr,
+                               /*treatUsableFromInlineAsPublic*/true);
+  return scope.isPublic();
+}
+
+static bool isPublicOrUsableFromInline(Type ty) {
+  // Note the double negative here: we're looking for any referenced decls that
+  // are *not* public-or-usableFromInline.
+  return !ty.findIf([](Type typePart) -> bool {
+    // FIXME: If we have an internal typealias for a non-internal type, we ought
+    // to be able to print it by desugaring.
+    if (auto *aliasTy = dyn_cast<NameAliasType>(typePart.getPointer()))
+      return !isPublicOrUsableFromInline(aliasTy->getDecl());
+    if (auto *nominal = typePart->getAnyNominal())
+      return !isPublicOrUsableFromInline(nominal);
+    return false;
+  });
+}
+
+namespace {
+  class InheritedProtocolCollector {
+    SmallVector<ProtocolDecl *, 8> IncludedProtocols;
+    SmallVector<ProtocolDecl *, 8> ExtraProtocols;
+
+    void recordProtocols(ArrayRef<TypeLoc> directlyInherited) {
+      for (TypeLoc inherited : directlyInherited) {
+        Type inheritedTy = inherited.getType();
+        if (!inheritedTy || !inheritedTy->isExistentialType())
+          continue;
+
+        bool canPrintNormally = isPublicOrUsableFromInline(inheritedTy);
+        SmallVectorImpl<ProtocolDecl *> &whichProtocols =
+            canPrintNormally ? IncludedProtocols : ExtraProtocols;
+
+        ExistentialLayout layout = inheritedTy->getExistentialLayout();
+        for (ProtocolType *protoTy : layout.getProtocols())
+          whichProtocols.push_back(protoTy->getDecl());
+      }
+    }
+
+  public:
+    using PerTypeMap = llvm::MapVector<const NominalTypeDecl *,
+                                       InheritedProtocolCollector>;
+
+    static void collectProtocols(PerTypeMap &map, const Decl *D) {
+      ArrayRef<TypeLoc> directlyInherited;
+      const NominalTypeDecl *nominal;
+      const IterableDeclContext *memberContext;
+
+      if ((nominal = dyn_cast<NominalTypeDecl>(D))) {
+        directlyInherited = nominal->getInherited();
+        memberContext = nominal;
+
+      } else if (auto *extension = dyn_cast<ExtensionDecl>(D)) {
+        if (extension->isConstrainedExtension()) {
+          // Conditional conformances never apply to inherited protocols, nor
+          // can they provide unconditional conformances that might be used in
+          // other extensions.
+          return;
+        }
+        nominal = extension->getExtendedNominal();
+        directlyInherited = extension->getInherited();
+        memberContext = extension;
+
+      } else {
+        return;
+      }
+
+      map[nominal].recordProtocols(directlyInherited);
+
+      // Recurse to find any nested types.
+      for (const Decl *member : memberContext->getMembers())
+        collectProtocols(map, member);
+    }
+
+    void
+    printSynthesizedExtensionIfNeeded(raw_ostream &out,
```

Comment:
Can you add a comment here explaining the motivation for why we're printing the empty extensions after the types? I think it's useful to mention that we are printing conformances to public types that are hidden by conformances to private types, and that they're all satisfied by the requirements that have already been printed.

Reply:
Oops. Yes, this whole scheme deserves commenting.

Reply:
It kind of stinks to print the extensions at the end of the file rather than next to the type, but I don't see a good way around that.